server/app_multi/app.py

```diff
@@ -0,0 +1,163 @@
+import os
+import datetime
+
+from flask import Flask, redirect, current_app, make_response, render_template
+from flask import Blueprint, request
+from flask_caching import Cache
+from flask_compress import Compress
+from flask_cors import CORS
+from flask_restful import Api, Resource
+
+from server.data_common.utils import Float32JSONEncoder, MatrixDataLoader
+
+from http import HTTPStatus
+from functools import wraps
+import server.common.rest as common_rest
+from server.common.utils import path_join
+
+web_bp = Blueprint("webapp", "server.common.web", template_folder="templates")
+
+
+@web_bp.route("/")
+def index():
+    # FIXME with a splash screen that includes a listing of all the datasets.
+    # or perhaps a login screen if this is a hosted environment
+    return "<H1>Welcome to cellxgene<H1>"
+
+
+def get_data_engine(dataset):
+    config = current_app.app_config
+    location = path_join(config.dataroot, dataset)
+    matrix_data_loader = MatrixDataLoader(location)
```

Comment:
This seems to assume that reloading the data will be cheap (it is done on every single HTTP request).   In fact, for libraries like AnnData, which assume all data is pre-loaded and retained in memory, it is very expensive.

This is another case where there is either going to be abstraction leakage, or you need to make the MatrixDataLoader more complicated by adding in-memory caching for the "multi-dataset, loading H5AD" use case.

Reply:
I addressed this by adding a cache so that data does not need to be reloaded between requests.